url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/common/CircularDoubleLinkedList.java

```diff
@@ -0,0 +1,261 @@
+    }
+
+    public void add(E e)
+    {
+        _CircularDoubleLinkedListNode<E> newContainer;
+        newContainer = new _CircularDoubleLinkedListNode<E>();
+        newContainer.data = e;
+        if ( head == null ) {
+            head = newContainer;
+            newContainer.next = newContainer;
+            newContainer.previous = newContainer;
+        }
+        else {
+            newContainer.previous = head.previous;
+            newContainer.next = head;
+            head.previous.next = newContainer;
+            head.previous = newContainer;
+        }
+        currentSize++;
+    }
+
+    public void insertBefore(E newElem, E pivot)
+    {
+        locateWindowAtElem(pivot);
+        _CircularDoubleLinkedListNode<E> newContainer;
+
+        newContainer = new _CircularDoubleLinkedListNode<E>();
+        newContainer.data = newElem;
+        lastAccessedIndex = -1;
+
+        if ( head == null ) {
+            head = newContainer;
+            newContainer.next = newContainer;
+            newContainer.previous = newContainer;
+        }
+        else if ( window == null || window == head ) {
+            window = head;
+            head = newContainer;
+            newContainer.previous = window.previous;
+            newContainer.next = window;
+            window.previous.next = newContainer;
+            window.previous = newContainer;
+        }
+        else {
+            newContainer.previous = window.previous;
+            newContainer.next = window;
+            window.previous.next = newContainer;
+            window.previous = newContainer;
+        }
+        currentSize++;
+    }
+
+    public void locateWindowAtIndex(int index)
+    {
+        if ( index < 0 || index >= currentSize ) {
+            return;
+        }
+        int i;
+        for ( i = 0, window = head;
+              i < currentSize && i < index;
+              i++, window = window.next );
+        lastAccessedIndex = i;
+    }
+
+    public boolean locateWindowAtElem(E e)
+    {
+        int i;
+
+        lastAccessedIndex = -1;
+        for ( i = 0, window = head;
+              i < currentSize;
+              i++, window = window.next ) {
+            if ( window.data == e ) {
+                lastAccessedIndex = i;
+                return true;
+            }
+        }
+        window = null;
+        return false;
+    }
+
+    public void swapElements(E e1, E e2)
+    {
+        locateWindowAtElem(e1);
+        _CircularDoubleLinkedListNode<E> window1 = window;
+        locateWindowAtElem(e2);
+        _CircularDoubleLinkedListNode<E> window2 = window;
+
+        if ( window1 == null || window2 == null ) return;
+        E temp = window1.data;
+
+        window1.data = window2.data;
+        window2.data = temp;
+    }
+
+    public E next()
+    {
+        lastAccessedIndex = -1;
+        if ( window == null ) {
+            window = head;
+        }
+        E elem = window.data;
+        window = window.next;
+        return elem;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 137:143, 158:164.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.